assembly/systemCalls.ts

```diff
@@ -768,27 +771,103 @@ export namespace System {
+    return result.value;
+  }
+
+  /**
+   * Legacy function to check authority for an account (not secure,
+   * it is recommended to use System.checkAuthority or
+   * System.checkCallContractAuthority)
+   * @param type type of authority required
+   * @param account account to check
+   * @param data data to be passed
+   * @returns bool true if the account has authority
+   * @example
+   * @deprecated
+   * ```ts
+   * System.checkAuthorityLegacy(authority.authorization_type.transaction_application, Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe));
+   * ```
+   */
+  export function checkAuthorityLegacy(type: authority.authorization_type, account: Uint8Array, data: Uint8Array | null = null): bool {
+    const args = new system_calls.check_authority_arguments(type, account, data !== null ? data : new Uint8Array(0));
+    const encodedArgs = Protobuf.encode(args, system_calls.check_authority_arguments.encode);
+
+    const retcode = env.invokeSystemCall(system_call_ids.system_call_id.check_authority, SYSTEM_CALL_BUFFER.dataStart as u32, MAX_BUFFER_SIZE, encodedArgs.dataStart as u32, encodedArgs.byteLength, RETURN_BYTES.dataStart as u32);
+    checkErrorCode(retcode, SYSTEM_CALL_BUFFER.slice(0, RETURN_BYTES[0]));
+    const result = Protobuf.decode<system_calls.check_authority_result>(SYSTEM_CALL_BUFFER, system_calls.check_authority_result.decode, RETURN_BYTES[0]);
+    return result.value;
+  }
+
+  /**
+   * Check authority for an account
+   * @param type type of authority required
+   * @param account account to check
+   * @param data data to be passed. By default it uses operation args
+   * @param caller contract caller. By default it calls the function to get the caller
+   * @returns bool true if the account has authority
+   * @example
+   * ```ts
+   * // check contract call authority
+   * const isAuthorized = System.checkAuthority(
+   *   authority.authorization_type.contract_call,
+   *   Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe
+   * );
+   *
+   * // you can also pass args and caller. By default it gets them automatically,
+   * // and they are cached to avoid multiple system calls
+   * System.checkAuthority(
+   *   authority.authorization_type.contract_call,
+   *   Base58.decode('1DQzuCcTKacbs9GGScRTU1Hc8BsyARTPqe),
+   *   args,
+   *   caller
+   * );
+   * ```
+   */
+  export function checkAuthority(
+    type: authority.authorization_type,
+    account: Uint8Array,
+    data: Uint8Array | null = getArguments().args,
+    caller: Uint8Array | null = getCaller().caller
+  ): bool {
+    // if there is a caller and the account does not use a
+    // smart wallet then reject the operation. Otherwise call
+    // the native check authority thunk
+    if (caller && caller.length > 0) {
+      if (Arrays.equal(caller, account)) return true;
+      const contractMetadata = getContractMetadata(account);
+      if (
+        !contractMetadata ||
```

Comment:
I'm not sure I understand why we return `false` if there is a `caller` and `account` does not have a smart contract. Couldn't we be checking the authority of an account that is not the `caller` and does not have a smart contract?

e.g.:
- `caller` is contract A
- I want to check the authority of `account` B (which does not have a smart contract)
- `checkAuthority` should call the native check authority thunk and not return false?

Reply:
After reading that post please read https://peakd.com/koinos/@jga/new-koinos-system-call-live-in-the-testnet which explains why we changed the approach to the `get_contract_metadata` system call (the proposal that was finally approved). But the initial motivation and functionality required remains the same.

Reply:
Got it! I think this will have to be explained somewhere because it's going to be confusing for devs at first. (_"Why is it returning false when the user did sign the transaction?"_)